src/dlit_lu/constraint.py

```diff
@@ -0,0 +1,443 @@
+
+        cagr_columns = []
+
+        for i in range(1, len(year_columns)):
+            start_year = year_columns[i - 1]
+            end_year = year_columns[i]
+            years = end_year - start_year
+
+            def calculate_row_growth(row):
+                start_value = row[str(start_year)]
+                end_value = row[str(end_year)]
+                if start_value == 0:
+                    return None  
+                return ((end_value / start_value) ** (1 / years) - 1) * 100
+
+            cagr_column_name = f'CAGR_{end_year}'
+            growth_rate[cagr_column_name] = data.apply(calculate_row_growth, axis=1)
+            cagr_columns.append(cagr_column_name)
+
+        result = growth_rate.iloc[:, [0, 1, 2] + [growth_rate.columns.get_loc(col) for col in cagr_columns]]
+
+        return result
+
+def run(config: inputs.DLitConfig):
+
+    if config.dev_pattern is None:
+        raise ValueError("Cannot run development pattern without any dev_pattern parameters")
+
+    LOG.info("Initialising GrowthRate Module")
+
+    config.output_folder.mkdir(exist_ok=True)
+
+    # Load data
+    ddg_pop = pd.read_csv(config.constraint.ddg_pop)
+    ddg_emp = pd.read_csv(config.constraint.ddg_emp)
+    dlog_population = pd.read_csv(config.constraint.dlog_population)
+    dlog_employment = pd.read_csv(config.constraint.dlog_employment)
+
+    key_constraint_path = config.output_folder / f"06_constraint_test"
+    key_constraint_path.mkdir(exist_ok=True)
+
+    # Process data
+    year_columns = [col for col in dlog_population.columns if col.isdigit()]
+    ddg_col = 'LAD13CD'
+    ddg_pop = ddg_pop[[ddg_col] + year_columns]
+    ddg_emp = ddg_emp[[ddg_col] + year_columns]
+
+    ddg_pop = ddg_pop[~ddg_pop[ddg_col].str.startswith('LON')]
+    ddg_emp = ddg_emp[~ddg_emp[ddg_col].str.startswith('LON')]
+
+    ddg_pop = ddg_pop.rename(columns={'LAD13CD': 'lad2013_id'})
+    ddg_emp = ddg_emp.rename(columns={'LAD13CD': 'lad2013_id'})
+
+    processor = ConstraintProcessor(config)
+    growth_calculator = GrowthCalculator()
+
+    lad_id = "lad2013_id"
+    name_column = "descriptions"
+    base_year_column = "2023"
+    base_year_int = int(base_year_column)
+    region_id = "ntem_region_id"
+    name_column_region = "zone_id"
+
+    build_out_columns = np.arange(base_year_int + 1, 2067, 1).tolist()
+    build_out_columns = [str(year) for year in build_out_columns]
+
+    # Aggregate data
+    region_dlog_pop = processor.region(
+        dlog_population, 
+        base_year_column, 
+        build_out_columns)
+    
+    region_ddg_pop = processor.region(
+        ddg_pop, 
+        base_year_column, 
+        build_out_columns)
+    
+    region_ddg_emp = processor.region(
+        ddg_emp, 
+        base_year_column, 
+        build_out_columns)
+    
+    region_dlog_emp = processor.region(
+        dlog_employment,
+        base_year_column, 
+        build_out_columns)
+
+    # Add names
+    datasets = [
```

Comment:
Could you not just call processor.add_names_to_data here instead?
```suggestion
    processed_datasets = [
        processor.add_names_to_data(ddg_pop, lad_id, name_column, False, False),
        processor.add_names_to_data(ddg_emp, lad_id, name_column, False, False),
        processor.add_names_to_data(dlog_population, lad_id, name_column, False, False),
        processor.add_names_to_data(dlog_employment, lad_id, name_column, False, False),
        processor.add_names_to_data(region_ddg_pop, region_id, name_column_region, True, True),
        processor.add_names_to_data(region_ddg_emp, region_id, name_column_region, True, True),
        processor.add_names_to_data(region_dlog_pop, region_id, name_column_region, True, True),
        processor.add_names_to_data(region_dlog_emp, region_id, name_column_region, True, True)
```

Reply:
Sorry that I did not make it clear-- the code in constraint module was still on-going and not ready for review.